Add scopeQuery for UserTask

guyyoo opened the pull request.

Hiro-Nakamura commented on handlers/userform-create.js:
```diff
@@ -135,6 +138,8 @@ module.exports = {
             req.log("created form:", form.uuid);
             cb(null, form);
 
+            if (newForm.scopeQuery) return;
```
So do we not broadcast new inbox items if they have a `scopeQuery` defined?  

Will they only see the inbox item when they refresh the browser / mobile app?

Bring this up with the team to see if that is the design we want.

However, if we want to send out broadcast updates on the correct users, then we need to:
- gather all `potentialUsers` = ( `users` + usersFromRoles(`roles`) )
- for each `potentialUsers`: if they pass the scopeQuery, then add them to the `listOfUpdateUsers`
- then perform `await req.broadcast.inboxCreate(listOfUpdateUsers, null, form);`

to do this, you should pull out the `isCorrectProcesForm()` logic out of `inbox-find` and put it in a shared `utility` that both `inbox-find` and `userform-create` can use.

nh758 replied:
> So do we not broadcast new inbox items if they have a `scopeQuery` defined?

I'm ok with this behavior short term. I'd like to get this useable first and then come back to this.

Actually we need to figure this out for our model updates as well. Currently we don't broadcast if the scope has inUser query. It might be good to tackle these together, soon.

Hiro-Nakamura reviewed (changes requested):
Some thoughts: 

You're using a nice concise `async` => `await` design pattern where you can chain embedded `async` calls.  

However, this design pattern also forces you to put some code in some of the embedded loops, that don't need to be performed on each iteration of the loop.

see my comment on the `users.map()` below.

Another example would be the definition of `const getDataQueries = async () =>` below.  It is an async function that could be defined 1x and used with the proper parameters.  However in the current code, it is being defined in the `filterConditions.rules.map(async (rule) => {` loop.  

That means for each `filterCondition` we are setting aside some memory for the definition of `getDataQueries()`, using it for that one loop iteration, then discarding it for the very next iteration where we will redefine another copy of that function.

This memory isn't freed up until the javascript garbage collector comes though and frees up this unused memory.  You can imagine that our memory usage will scale exponentially if we receive N requests with M loops each request.  

So lets refactor this so we don't waste so much memory when we don't have to.

guyyoo commented:
> Some thoughts:
> 
> You're using a nice concise `async` => `await` design pattern where you can chain embedded `async` calls.
> 
> However, this design pattern also forces you to put some code in some of the embedded loops, that don't need to be performed on each iteration of the loop.
> 
> see my comment on the `users.map()` below.
> 
> Another example would be the definition of `const getDataQueries = async () =>` below. It is an async function that could be defined 1x and used with the proper parameters. However in the current code, it is being defined in the `filterConditions.rules.map(async (rule) => {` loop.
> 
> That means for each `filterCondition` we are setting aside some memory for the definition of `getDataQueries()`, using it for that one loop iteration, then discarding it for the very next iteration where we will redefine another copy of that function.
> 
> This memory isn't freed up until the javascript garbage collector comes though and frees up this unused memory. You can imagine that our memory usage will scale exponentially if we receive N requests with M loops each request.
> 
> So lets refactor this so we don't waste so much memory when we don't have to.

Thank you for bringing this to my attention. I have reviewed and addressed the issues. However, please feel free to verify and confirm that the changes have been successfully implemented.

Hiro-Nakamura commented on handlers/inbox-find.js:
```diff
@@ -45,94 +45,266 @@ module.exports = {
+
+                  try {
+                     const dataProcessInstance = (
+                        await AB.objectProcessInstance()
```
Any of these calls too `.model().find()`  should be wrapped in a `req.retry()` to make sure we can handle common DB connection problems.

So, for example:
```
await req.retry(() => AB.objectProcessInstance().model().find({ uuid: l.process }, req))[0]
```

Hiro-Nakamura commented on handlers/inbox-find.js:
```diff
@@ -45,94 +45,266 @@ module.exports = {
+                        await Promise.all(
+                           users.map(async (u) => {
+                              const user = { username: u };
+                              const data = await query
```
req.retry()

Hiro-Nakamura reviewed (changes requested):
Good changes.

Also, make sure we wrap DB operations in a `req.retry()` call.

guyyoo commented:
> Good changes.
> 
> Also, make sure we wrap DB operations in a `req.retry()` call.

Ok, the new fixes is on there.

Hiro-Nakamura reviewed (approved):
good job!
